Tabulated Fluid Properties support for (v,e) set, rebased

licharlot opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

Commit:
Allow TabulatedFluidProperties to evaluate x_from_v_e functions

Commit:
Testing fluid property lookup in terms of v and e

Commit:
Enable computing e_from_v_h

Commit:
Implement AD interfaces (#20101)

spaubt2 committed:
mplement Bilinear Interpolation Table For Fluid Properties refs #20101

spaubt2 committed:
Use TabulatedFluidProperties base class for bilinear and bicubic

spaubt2 committed:
Fix compilation warnings wrt initialization of temporatry variables, fix docstrings for fluid property routines in TFP, adjust header selection in source file for TFP
refs #20101

spaubt2 committed:
rewrote header and source files of Bi*Interpolation to have syntax match eachother for Bidimensional, Bilinear, and Bicubic. Allowed for adaptation of inheritance system where Bidimensional is the parent class of the other two.
refs #20101

spaubt2 committed:
New names for routines specific to interpolation type oin TabulatedFP
Refs #20101

spaubt2 committed:
Update tabulatedFP documentation to new ve & linear capability, refs #20101

spaubt2 committed:
Add regression testing to new Tabuleated ve linear capability, refs #20101

spaubt2 committed:
Add routines to TabulatedFP for T_from_prho s_from_ve pt_from_hs
refs #20101

spaubt2 committed:
Added routines to SinglePhaseFluidProperties using NewtonMethod to go from v,e to p,T. These routines get used in TFP as well. Modified TabulatedBicubicFluidProperties and TabulatedBilinearFluidProperties to change how interpolate data to change variable sets. Now uses 2D Newton Method to convert between variable sets. TFP WIP still ref #20101

spaubt2 committed:
Created util that does NewtonMethod in 1D and 2D in the fluid_properties module. ref #20101

spaubt2 committed:
fixed e_from_p_rho to get the correct derivatives. Implemented derivatives for T(p,rho) using finite difference method. Added feature to TBicubicFP to allow user to choose whether to get p and T from v,e or v,h or choose to use both.

spaubt2 committed:
Modifed TabulatedBilinearFluidProperties to include ability to choose between from_ve, from_vh, or use both to build tabulated data ref #20101

spaubt2 committed:
Added checkNaNs routine to TabBicubicFP and TabBilinearFP. If nans exist, set variable to be constant as min or max, depending on where nans exist. Also added a checkOutofBounds routine to set variable to be constant at min or max depending on where variable is out of bounds. Added T_from_p_h routine using Newton Method. Added s_from_p_h routine which uses T_from_p_h, then finds s_from_p_T. Added feature to TabBicubicFP and TabBilinearFP which allows user to choose to generate (p,T) from both (v,e) and (v,h) or choose one or the other. Added routine to TFP to check the initial guesses for p and T that are used in Newton's Mehotd. If these guesses are outside the range for max/min of p and T, routine produces an error and informs the user. Ref #20101

spaubt2 committed:
changed fp to a pointer. This allowed for the ability to choose not to generate fluid properties from _fp. ref #20101

spaubt2 committed:
Format function declarations

spaubt2 committed:
Fixed bug in TFP where if statement for constructing pT_from_ve missing ../../moose/modules/navier_stokes/navier_stokes-devel -i functorfluidpropsVE.i Edited documentation in TFP.md file ref #20101

spaubt2 committed:
Added FluidPropertiesMaterialPH and FluidPropertiesMaterialVE. When used with input files, these allow user to get properties from (v,e) and (p,h). ref #20101

spaubt2 committed:
Resolved issue where fluid properties are initialized twice, reading the csv file twice, causing them to error.

TFP can now run without dependence on another fp object, a csv file with data is all that is needed for this method to work

When variable set inversion of (p,T) from (v,e) or (v,h) occurs while using csv file, it is possible there will be (p,T) values that are out of bounds (converting from a (v,e) or (v,h) grid). When this is the case, pressure and temperature are reset to their respective range min or max values depending on the situation.

ref #20101

spaubt2 committed:
modified outputWarnings routine in TabulatedBi*FluidProperties.

Now outputs all warnings during variable set inversion at once to provide cleaner look.

ref # 20101

spaubt2 committed:
Resolved issue where fluid properties are initialized twice, reading the csv file twice, causing them to error.

spaubt2 committed:
Modified checkNaNs routine.

Previously was setting temperature and pressure values in the same way based on the specific volume. Should be set in opposite ways due to relation to volume.

ref #20101

spaubt2 committed:
Updated TabulatedFluidProperties.md file to document changes made to variable set inversions that now use Newton's Method, as well as implementaion of a few new routines from older commits. (e_from_p_rho, T_from_p_rho, etc.) ref #20101

spaubt2 committed:
Worked on passing regression tests with TFP.

First time doing this with new capabilities since ___.

WIP, bilinear interpolation is having some issues with getting NaNs in sampleValueandDerivatives routine.

ref #20101

spaubt2 committed:
Worked on implementing SampleValueAndDerivatives to BilinearInterpolation. Lots of special cases were needed. WIP ref #20101

spaubt2 committed:
Worked on fixing derivatives for BilinearInterpolation WIP ref#20101

spaubt2 committed:
created unit tests for BilinearInterpolation utility and NewtonInversion utility WIP ref #20101

spaubt2 committed:
Worked on unit test for BilinearInterpolation and NewtonInversion ref #20101

spaubt2 committed:
Modified BilnearInterpolation.C to obtain derivatives when x1=x2 and y1=y2 along the tob and bottom boundaries

ref #20101

spaubt2 committed:
Removed comments to console used while developing sampleDerivative routine

ref #20101

licharlot committed:
Removing commented lines and clang format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1167019/) on 1c8f53d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format caa42b12ae7bd296d356d59fadd112c0540f6bc5`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1167059/) on 3a922c3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22060/site)

This comment will be updated on new commits.

GiudGiud committed:
Finish newton inversion tests, refs #20101
for:
- y = f(x,z) problems, x is constant, search for z knowing y
- (y1,y2)=(f(x1, x2), g(x1, x2)), y1,y2 known, search for x1 and x2
These two problems arise during variable set inversion in fluid properties objects

GiudGiud committed:
Add NaN checks in Newton method solves, refs #20101

GiudGiud committed:
Fix unity build by adding bidimensional interpolation header, refs #20101

GiudGiud committed:
Add documentation to the fluid properties materials, refs #20101

GiudGiud committed:
Remove seg fault from unit test for bilinear interpolation

GiudGiud committed:
Fix bilinear interpolation unit test

GiudGiud committed:
Add left and right boundary nodes special treatment in
bilinear tabulation
Add coments for which region is being considered
Fix manual indexing with correct index sizes

GiudGiud committed:
Add exact grid interior node special treatment to bilinear interpolation

GiudGiud committed:
Cover all cases with unit tests for bilinear interpolation

GiudGiud committed:
Fix external app failures: define T_h_p, resiliently create FP_uos when params are set hastily

GiudGiud committed:
Fix unit testing for tabulated fluid properties, except linear FP

GiudGiud committed:
Improve docstrings in TFP
Improve parameter descriptions in SFP

GiudGiud committed:
Remove tabulated bilinear FP for investigation

GiudGiud committed:
Revert test to previous state

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1174027/) on 2ca3086 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b79b920ccf2f306cb40ff485f11c6b3c07de50ea`

GiudGiud committed:
Clang format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1174064/) on b7610c9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b79b920ccf2f306cb40ff485f11c6b3c07de50ea`

GiudGiud committed:
Regold ve tabulation due to change from Brent to Newton method for variable set inversion, refs #20101

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1174133/) on 828e76e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b79b92</th>
      <th colspan=4>#22060 828e76</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.17%</td>
      <td>84.17%</td>
      <td>+0.00%</td>
      <td>84.16%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>79112</td>
      <td>79419</td>
      <td>+307</td>
      <td>324</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14875</td>
      <td>14932</td>
      <td>+57</td>
      <td>61</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22060/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22060/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>b79b92</th>
      <th colspan=4>#22060 828e76</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.44%</td>
      <td>83.97%</td>
      <td>-1.46%</td>
      <td>67.85%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>5643</td>
      <td>6098</td>
      <td>+455</td>
      <td>498</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>962</td>
      <td>1164</td>
      <td>+202</td>
      <td>236</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22060/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22060/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22060/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 84.16% is less than the suggested 90.0%
- `fluid_properties` new line coverage rate 67.85% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1174133/) on 828e76e wanted to post the following:


The following coverage requirement(s) failed:
- `fluid_properties` coverage rate 83.97% is less than the required 84.0%

GiudGiud commented:
I ll work on getting coverage up.

GiudGiud committed:
Remove deprecated libmesh routine in bicubic interpolation, refs #20101

GiudGiud committed:
Add more unit tests to increase framework coverage of bi-dimensional interpolations, refs #20101

GiudGiud commented:
Coverage should be good in framework now. 
@lindsayad the part in the framework folder should be ready for review. 

Working on FP module coverage, and adding more exceptions and warnings.

GiudGiud commented:
```
##########################################################################
ERROR: The following files do not contain a newline character before EOF:
	unit/src/BilinearInterpolationTest.C

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################
```
never had that one. precheck is such a waste of time. just format my PR after it s merged..

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1174255/) on abfa09d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b79b920ccf2f306cb40ff485f11c6b3c07de50ea`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1174324/) on 5d4f5c8 wanted to post the following:


The following coverage requirement(s) failed:
- `fluid_properties` coverage rate 83.97% is less than the required 84.0%

GiudGiud committed:
Cleanup fluid property materials, which are used to output fluid properties using the various formulations

lindsayad commented on framework/include/timesteppers/TimeStepper.h:
```diff
@@ -136,7 +136,7 @@ class TimeStepper : public MooseObject, public Restartable, public ScalarCouplea
   Real & _timestep_tolerance;
 
-  ///should detailed diagnostic output be printed
+  /// should detailed diagnostic output be printed
```
```suggestion
  /// whether detailed diagnostic output should be printed
```

lindsayad commented on framework/include/utils/BicubicInterpolation.h:
```diff
@@ -10,6 +10,9 @@
 
 #include "MooseTypes.h"
+#include "BidimensionalInterpolation.h"
+
+class BidimensionalInterpolation;
```
```suggestion
```

lindsayad commented on framework/include/utils/BidimensionalInterpolation.h:
```diff
@@ -0,0 +1,88 @@
+
+// MOOSE includes
+#include "ColumnMajorMatrix.h"
+
```
```suggestion
```

lindsayad commented on framework/include/utils/BidimensionalInterpolation.h:
```diff
@@ -0,0 +1,88 @@
+ * This class interpolates tabulated data with a Bidimension function (either bicubic or bilinear).
+ * In order to minimize the computational expense of each sample, the coefficients at each point in
+ * the tabulated data are computed once in advance, and then accessed during the interpolation.
+ *
```
```suggestion
```

lindsayad commented on framework/include/utils/BidimensionalInterpolation.h:
```diff
@@ -0,0 +1,88 @@
+  /**
+   * Samples first derivative at point (x1, x2)
+   */
+  virtual Real sampleDerivative(Real /*x1*/, Real /*x2*/, unsigned int /*deriv_var*/) const
+  {
+    mooseError("sampleDerivative is not implemented for this interpolation class");
+  }
+
+  /**
+   * Samples second derivative at point (x1, x2)
+   */
+  virtual Real sample2ndDerivative(Real /*x1*/, Real /*x2*/, unsigned int /*deriv_var*/) const
+  {
+    mooseError("sample2ndDerivative is not implemented for this interpolation class");
+  }
+
+  /**
+   * Samples value and first derivatives at point (x1, x2)
+   * Use this function for speed when computing both value and derivatives,
+   * as it minimizes the amount of time spent locating the point in the
+   * tabulated data
+   */
+  virtual void sampleValueAndDerivatives(
+      Real /*x1*/, Real /*x2*/, Real & /* y*/, Real & /* dy1*/, Real & /* dy2*/) const
+  {
+    mooseError("sampleValueAndDerivatives is not implemented for this interpolation class");
+  }
+
+  /**
+   * Same as sampleValueAndDerivatives, but ADReal instead of Real
+   */
+  virtual void ADsampleValueAndDerivatives(const ADReal & /*x1*/,
+                                           const ADReal & /*x2*/,
+                                           ADReal & /* y*/,
+                                           ADReal & /* dy1*/,
+                                           ADReal & /* dy2*/) const
+  {
+    mooseError("sampleValueAndDerivatives is not implemented for this interpolation class");
+  }
```
I take it that there are some derived classes that don't implement these? If all derived classes implement a given function, I would much prefer these to be pure virtual

GiudGiud replied:
yes it doesnt make sense for all to be implemented

lindsayad commented on framework/include/utils/BidimensionalInterpolation.h:
```diff
@@ -0,0 +1,88 @@
+  /**
+   * Same as sampleValueAndDerivatives, but ADReal instead of Real
+   */
+  virtual void ADsampleValueAndDerivatives(const ADReal & /*x1*/,
```
```suggestion
  virtual void sampleValueAndDerivatives(const ADReal & /*x1*/,
```
No reason not to overload this one as we've done with the others. Moreover, this name violates our method naming style (with the beginning capital letter)

lindsayad commented on framework/include/utils/BilinearInterpolation.h:
```diff
@@ -11,18 +11,21 @@
 // C++ includes
 #include <vector>
 
+class BidimensionalInterpolation;
+
```
```suggestion\
```

GiudGiud replied:
uh?

lindsayad replied:
lol. Remove the forward declaration is what I was after

GiudGiud replied:
ok done

lindsayad commented on framework/include/utils/BilinearInterpolation.h:
```diff
@@ -52,15 +55,30 @@ class BilinearInterpolation
-  std::vector<Real> _yAxis;
+  /// sampleInternal only used by BilinearInterpolation, hence made private
+  template <typename T>
+  T sampleInternal(T & s1, T & s2) const;
```
```suggestion
  T sampleInternal(const T & s1, const T & s2) const;
```

lindsayad commented on framework/src/utils/BicubicInterpolation.C:
```diff
@@ -53,26 +52,23 @@ BicubicInterpolation::sample(const ADReal & x1, const ADReal & x2) const
-T
-BicubicInterpolation::sampleInternal(T x1, T x2) const
+template <class C>
+C
```
Was this to get credit for more lines? :laughing: `T` is kind of the canonical template parameter. I don't see that `C` is more explanatory. Anyway I don't really care about this, but in general I prefer to use `T` instead of `C`

GiudGiud replied:
C was chosen for a reason because T was failing actually. 
@spaubt2 do you remember why?

lindsayad replied:
I could see this as an issue if it was shadowing a class template parameter but I don't see how that would be the case here. Anyway this is not that important of a mystery to solve

lindsayad commented on framework/src/utils/BicubicInterpolation.C:
```diff
@@ -337,16 +371,33 @@ BicubicInterpolation::tableDerivatives(std::vector<std::vector<Real>> & dy_dx1,
   mooseAssert(x.size() >= 2,
               "There must be at least two points in the table in BicubicInterpolation");
 
+  // if (xi <= x[0])
+  // {
+  //   mooseDoOnce(mooseWarning("Bicubic interpolation sampled out of bounds"));
+  //   klo = 0;
+  //   khi = 1;
+  //   xs = 0;
+  //   return;
+  // }
+  // else if (xi >= x.back())
+  // {
+  //   mooseDoOnce(mooseWarning("Bicubic interpolation sampled out of bounds"));
+  //   klo = x.size() - 2;
+  //   khi = x.size() - 1;
+  //   xs = 1;
+  //   return;
+  // }
+
```
```suggestion
```

lindsayad commented on framework/src/utils/BidimensionalInterpolation.C:
```diff
@@ -0,0 +1,21 @@
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "BidimensionalInterpolation.h"
+#include "MooseError.h"
+#include "MathUtils.h"
+#include "MooseUtils.h"
+
+#include "libmesh/int_range.h"
```
```suggestion
```

lindsayad commented on framework/src/utils/BilinearInterpolation.C:
```diff
@@ -56,18 +56,30 @@ BilinearInterpolation::getNeighborIndices(const std::vector<Real> & inArr,
 template <typename T>
 T
-BilinearInterpolation::sample(const T & xcoord, const T & ycoord) const
+BilinearInterpolation::sampleInternal(T & s1, T & s2) const
```
```suggestion
BilinearInterpolation::sampleInternal(const T & s1, const T & s2) const
```
we're not modifying these are we? Would be a little concerning if we are

GiudGiud replied:
yes that would be an issue

lindsayad commented on framework/src/utils/BilinearInterpolation.C:
```diff
@@ -76,32 +88,498 @@ BilinearInterpolation::sample(const T & xcoord, const T & ycoord) const
-  // if xcoord lies exactly on an xAxis node do linear interpolation
+  // if x1 lies exactly on an xAxis node do linear interpolation
   if (lx == ux)
+  {
     return fQ11 + (fQ12 - fQ11) * (y - y1) / (y2 - y1);
+  }
```
```suggestion
    return fQ11 + (fQ12 - fQ11) * (y - y1) / (y2 - y1);
```

lindsayad commented on framework/src/utils/BilinearInterpolation.C:
```diff
@@ -76,32 +88,498 @@ BilinearInterpolation::sample(const T & xcoord, const T & ycoord) const
-
+  // if x2 lies exactly on an yAxis node do linear interpolation
   if (ly == uy)
+  {
     return fQ11 + (fQ21 - fQ11) * (x - x1) / (x2 - x1);
+  }
```
```suggestion
    return fQ11 + (fQ21 - fQ11) * (x - x1) / (x2 - x1);
```

lindsayad commented on framework/src/utils/BilinearInterpolation.C:
```diff
@@ -76,32 +88,498 @@ BilinearInterpolation::sample(const T & xcoord, const T & ycoord) const
-template Real BilinearInterpolation::sample<Real>(const Real &, const Real &) const;
-template ADReal BilinearInterpolation::sample<ADReal>(const ADReal &, const ADReal &) const;
+Real
+BilinearInterpolation::sampleDerivative(Real s1, Real s2, unsigned int deriv_var) const
```
```suggestion
BilinearInterpolation::sampleDerivative(const Real s1, const Real s2, const unsigned int deriv_var) const
```

lindsayad commented on framework/src/utils/BilinearInterpolation.C:
```diff
@@ -76,32 +88,498 @@ BilinearInterpolation::sample(const T & xcoord, const T & ycoord) const
+    }
+  }
+  mooseError("Bilinear interpolation failed to select a case for x1= ", s1, " x2= ", s2);
+}
```
This function is massive. Is it possible to simplify this somehow? Also, could this be applied for AD as well?

GiudGiud replied:
this is to compute derivative so typically we dont use it with AD (unless we want the second derivative)

currently not simplifiable.
if we change from an average slope to a bigger box slope for the "on the grid line" case, then we can move the computation of dfdy etc outside of the ifs, keeping only the index selection in the if statements

lindsayad replied:
kk. I'll consider this resolved then

GiudGiud committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1174927/) on 7bd9bc6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b79b920ccf2f306cb40ff485f11c6b3c07de50ea`

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedBicubicFluidProperties.md:
```diff
@@ -0,0 +1 @@
+TabulatedFluidProperties.md
```
This is strange. Derived classes should have their own pages. I recommend linking to a base class page or using template file(s) to get what you want.

GiudGiud replied:
ok made it own class. There s really not much that is different from the base to the derived in terms of user-docs

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedFluidProperties.md:
````diff
@@ -94,10 +111,11 @@ divided into 50 and 100 equal points, respectively, then the input file syntax n
   []
 []
 ```
+Note that Bi___ should be replaced with Bicubic or Bilinear depending on desired the Interpolation method
````
- Blank line needed after code block
- Quotes around "Bi___", "Bicubic", and "Bilinear"
- on the desired
- lowercase "interpolation"
- Period

joshuahansel commented on modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h:
```diff
@@ -103,12 +105,31 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
 
   virtual void rho_from_p_T(
       Real pressure, Real temperature, Real & rho, Real & drho_dp, Real & drho_dT) const override;
+  virtual void rho_from_p_T(const ADReal & pressure,
+                            const ADReal & temperature,
+                            ADReal & rho,
+                            ADReal & drho_dp,
+                            ADReal & drho_dT) const override;
```
Are people using these interfaces with the AD derivatives?

GiudGiud replied:
good point, probably not. Let s delete it for now

GiudGiud replied:
they do. 
see in NSFV:
```
  addFunctorProperty<ADReal>(
      NS::time_deriv(NS::density),
      [this](const auto & r, const auto & t) -> ADReal
      {
        ADReal rho, drho_dp, drho_dT;
        _fluid.rho_from_p_T(_pressure(r, t), _temperature(r, t), rho, drho_dp, drho_dT);
        return drho_dp * _pressure.dot(r, t) + drho_dT * _temperature.dot(r, t);
      });
```

joshuahansel commented on modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h:
```diff
@@ -155,87 +214,130 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
-                     unsigned int ncol,
-                     const std::vector<Real> & vec,
-                     std::vector<std::vector<Real>> & mat);
+  Real inverseDistance(const std::vector<Real> & value, const std::vector<Real> & distance) const;
```
doxygen

GiudGiud replied:
removed routine

joshuahansel commented on modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h:
```diff
@@ -155,87 +214,130 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
 
-  /// File name of tabulated data file
+  // File name of tabulated data file
   FileName _file_name;
```
Don't unfix things

joshuahansel commented on modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h:
```diff
@@ -155,87 +214,130 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
+  bool _construct_pT_from_ve;
+  // if the lookup table p(v, h) and T(v, h) should be constructed
+  bool _construct_pT_from_vh;
+  // keeps track of whether initialSetup has been preformed
```
performed

joshuahansel commented on modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h:
```diff
@@ -155,87 +214,130 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
+  unsigned int _num_v;
+  // Number of internal energy points in tabulated data
+  unsigned int _num_e;
+  // to error or not on out of bounds check
```
out-of-bounds

joshuahansel commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -0,0 +1,64 @@
+// using 2D Newton Solve will need to write routines to take care of values where this
+// occurs. See TabulatedBicubicFluidProperties for example (routine checkOutofBounds).
+
+namespace NewtonMethod
```
I think a fluid properties namespace would be better: `FluidProperties` or `FluidPropertiesUtils`

joshuahansel commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -0,0 +1,64 @@
+#include <functional>
+
+// This file performs a Newton solve in 1D and 2D. In 1D, solve fails to converge
+// and error outputs if it takes more than 100 iterations. In 2D, if more than 100
```
I suggest making the max iterations a parameter to these functions

GiudGiud replied:
done

joshuahansel commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -0,0 +1,64 @@
+// and error outputs if it takes more than 100 iterations. In 2D, if more than 100
+// iterations occur, save results references and set convergence to false. Applications
+// using 2D Newton Solve will need to write routines to take care of values where this
+// occurs. See TabulatedBicubicFluidProperties for example (routine checkOutofBounds).
```
This file should get a corresponding doc file. You can refer to it on your other doc pages.

joshuahansel commented on modules/fluid_properties/src/utils/NewtonInversion.C:
```diff
@@ -0,0 +1,134 @@
+// and error outputs if it takes more than 100 iterations. In 2D, if more than 100
+// iterations occur, save results references and set convergence to false. Applications
+// using 2D Newton Solve will need to write routines to take care of values where this
+// occurs. See TabulatedBicubicFluidProperties for example (routine checkOutofBounds).
```
Don't need this text in both files

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedFluidProperties.md:
```diff
@@ -2,8 +2,9 @@
 
 !syntax description /Modules/FluidProperties/TabulatedFluidProperties
```
This page needs a good look-over.

GiudGiud replied:
done

joshuahansel commented:
I'll have more review comments after this round is addressed.

GiudGiud committed:
Add test for fluid properties materials, refs #20101

GiudGiud commented on modules/fluid_properties/include/materials/FluidPropertiesMaterialPH.h:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Computes fluid properties using (pressure, temperature) formulation
+ */
+class FluidPropertiesMaterialPH : public Material
```
I added the tests.
It s to be able to access these properties from the input file.
I could go either way, but there's already one checked in the repo as a normal object I think.

GiudGiud replied:
yeah Ben only added PH. PT and VE (without suffix) were already in the repo
I think they're convenient

joshuahansel replied:
Ok, it looks like these are intended for output only, not in any residual objects, since there are no derivatives.

- I think the VE and PT ones make some sense, since these variables are often solution variables. PH, not so much - does anyone actually want to use this?
- If these are used for output, then I say: why stop at the quantities that are directly computable by that variable set? I'd think the objective is to compute everything that you can (all of the quantities that `SinglePhaseFluidProperties` offers. Right now I notice `FluidPropertiesMaterialVE` and `FluidPropertiesMaterialPT` compute different sets of quantities, for example.

joshuahansel replied:
Note that `FluidPropertiesInterrogator` is doing the same thing as these materials, but of course these materials have the advantage of being able to do it at QPs rather than a single point. It'd probably be a nice refactor at some point to have some kind of thermodynamic state definition class (where all quantities would be computed, given some variable set), but I definitely won't ask for that here.

GiudGiud replied:
I ll move PH to a test object. 
I ll see if we can add more, but there's already a lot I would not be surprised if it s already maxed out

joshuahansel replied:
You have `T = T(p,h)` as your first call, so then you have `(p,T)` and thus everything. However, it's not really useful to add more things unless it's a regular (non-test) object.

GiudGiud replied:
well I like to be able to see the variation, but realistically I dont know who works fully with (p, h) and would require this. We only have 2 calls which are (p,h). 

I m thinking we dont actually want too many properties too. The more properties we have the more likely these materials may not be used with a fluid property UO, because of an unimplemented property

GiudGiud committed:
Apply review from Josh
- add Newton util documentation and change namespace
- remove unused routine
- fix up TFP documentation
Refs #20101

GiudGiud committed:
Buff TFP documentation, explain pitfalls of inversions between varibale sets
add Bicubic TFP documentation page
Clean up Newton inversion routines
Add max iteration parameter
refs #20101

GiudGiud commented:
@joshuahansel review should be addressed now

GiudGiud committed:
Emit a warning when the variable set conversion fails to converge, warning is subpar because of #17896

GiudGiud committed:
Add more exception checking, refs #20101

joshuahansel reviewed (changes requested):
More reviews to follow...

GiudGiud commented on modules/fluid_properties/doc/content/syntax/Utils/FluidPropertiesUtils.md:
```diff
@@ -0,0 +1,45 @@
+# Utilities for the Fluid Properties Module
```
didnt really know where to put it
but where do I say we have Brent s method too?

GiudGiud replied:
I dont want this to show up in moose source page.
This is specific to fluid properties module, but people want Newton solvers everywhere

GiudGiud replied:
this rules out content/source/utils.
content/utils should be ok?

joshuahansel replied:
> I dont want this to show up in moose source page.
This is specific to fluid properties module, but people want Newton solvers everywhere

I think it's fine for it to show up. It's a deficiency of the doc page that it puts everything together. And no one uses the source page lookup anyway :) It's just a logical place to organize documentation for source files, to be referenced in places where people actually go.

joshuahansel replied:
(But I don't care that much - put it where you wanna)

GiudGiud commented on modules/fluid_properties/doc/content/syntax/Utils/FluidPropertiesUtils.md:
```diff
@@ -0,0 +1,45 @@
+
+The 1D Newton method solver is used to solve the problem:
+
+- Find z such that $f(x, z) = y$, with $x$, $y$ known constants and $f$ a function. 
```
lolol use code suggestions

GiudGiud commented on modules/fluid_properties/include/userobjects/SinglePhaseFluidProperties.h:
```diff
@@ -328,6 +328,71 @@ class SinglePhaseFluidProperties : public FluidProperties
+   * @param[out] fluid pressure (Pa / kg)
+   * @param[out] Temperature (K)
+   */
+  virtual void p_T_from_v_e(const Real & v,
```
there's nothing specific about wanting to use maths to perform an inversion. 
in the vast majority of fluid properties, these conversions are not analytic, which is also why support for (v,e) is so limited outside of the FP repos.

GiudGiud replied:
if we dont have this here, we're going to have the same thing in all the downstream classes that are not over-simplified

joshuahansel replied:
My main reservation is that users of all fluid properties would now have the parameters "tolerance", "T_initial_guess", and "p_initial_guess", which are usually irrelevant for them, which is confusing to the user (do I need to set this? what does this do?). Also developers may use one of these solve methods, when they have more direct and efficient options to them. However, I'm fine with keeping this. I originally had thought this was used only in `TabulatedFluidProperties`, but now I see the new default implementations (which I'm guessing are needed indirectly in `TabulatedFluidProperties`).

GiudGiud replied:
those parameters are in the `Variable set conversions Newton solve` group.
I ll add docs in SFP.md

afaik there are no closed forms to convert between variable sets in the non-easy direction. The developers will be happy to find these routines rather than implement their own.

GiudGiud replied:
ugh we dont have that page. I ll upload all other revisions and work on writing it

GiudGiud commented on modules/fluid_properties/include/userobjects/TabulatedBicubicFluidProperties.h:
```diff
@@ -0,0 +1,150 @@
+ *
+ * Properties specified in the data file or listed in the input file (and their derivatives
+ * wrt pressure and temperature) will be calculated using bicubic interpolation, while all
+ * remaining fluid properties are calculated using the supplied FluidProperties UserObject.
```
removed

joshuahansel reviewed (changes requested):
More comments

joshuahansel commented:
I never realized when you've started another review, all of your responses to existing threads also become "pending"  - seems dumb...

GiudGiud commented on modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:
```diff
@@ -118,7 +146,8 @@ TabulatedFluidProperties::initialSetup()
       if (std::find(column_names.begin(), column_names.end(), _required_columns[i]) ==
           column_names.end())
-        mooseError("no ",
+        mooseError(name(),
```
fixed

GiudGiud commented on modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:
```diff
@@ -354,7 +470,70 @@ TabulatedFluidProperties::e_from_p_T(
+  Real T = FluidPropertiesUtils::NewtonSolve(pressure, rho, _T_initial_guess, _tolerance, lambda);
+  // check for nans
+  if (std::isnan(T))
+    mooseError("Conversion from pressure (p = ",
```
p_rho routines are for subchannel. yup time will tell!

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedFluidProperties.md:
```diff
@@ -21,17 +18,22 @@ Using an interpolation of tabulated fluid properties can significantly reduce th
-the fluid properties that TabulatedFluidProperties understands. These are: *density*, *enthalpy*,
-*internal_energy*, *viscosity*, *k* (the thermal conductivity), *cp* (the isobaric specific heat
-capacity), *cv* (the isochoric specific heat capacity), and *entropy*.
+the fluid properties that TabulatedFluidProperties understands. These are: *density*, *enthalpy*, *v* (specific_volume),
```
table or list

GiudGiud replied:
listed and fixed format a little

GiudGiud commented on modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:
```diff
@@ -474,31 +762,335 @@ TabulatedFluidProperties::s_from_p_T(Real pressure, Real temperature) const
+{
+  if (!_construct_pT_from_ve)
+    mooseError("You must construct pT from ve tables when calling p_from_v_e.");
+  Real x = 0;
+  Real raw1 = v.value();
+  Real raw2 = e.value();
+  Real dxd1 = 0;
+  Real dxd2 = 0;
+  p_from_v_e(raw1, raw2, x, dxd1, dxd2);
+
+  DualReal result = x;
+  result.derivatives() = v.derivatives() * dxd1 + e.derivatives() * dxd2;
+  return result;
```
seems so. nice catch

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedFluidProperties.md:
```diff
@@ -60,13 +62,20 @@ and so on.
 Consider the example where TabulatedFluidProperties is used to reduce the cost of calculating
```
Avoid referring to the class name since it's a base class name

joshuahansel replied:
Actually, I think it'd be ok to do: a `TabulatedFluidProperties` object

GiudGiud replied:
fixed

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedFluidProperties.md:
```diff
@@ -21,17 +18,22 @@ Using an interpolation of tabulated fluid properties can significantly reduce th
+
 ## File format
 
 The expected file format for the tabulated fluid properties is now described.  The first line must be
```
Refer to the corresponding parameter (via `[!param]`)

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedFluidProperties.md:
```diff
@@ -21,17 +18,22 @@ Using an interpolation of tabulated fluid properties can significantly reduce th
+The order is not important, although having pressure and temperature first makes the data easier for a human to read.
 
 The data in the pressure and temperature columns *must* be monotonically increasing. This file format
 does require duplication of the pressure and temperature data - each pressure value must be included
```
Not your changes, but it'd be nice to use symbols instead of "num_T" and "num_p"

GiudGiud committed:
Address Josh's review:
- move FPMat PH to a test object
- remove unecessary AD TFP routines which were stolent from SFP anyway
- remove name() from error calls as it s output anyway
- clean up TFP_ve test
- move util page out of syntax
Refs #20101

GiudGiud commented:
@joshuahansel should be addressed now. 
Working on an SFP doc page in case we dont have one (need to look)

joshuahansel commented on modules/fluid_properties/doc/content/source/userobjects/TabulatedFluidProperties.md:
```diff
@@ -109,8 +123,70 @@ FluidProperties UserObject.
-respect to pressure and temperature) will be calculated using bicubic interpolation, while all
-remaining fluid properties will be calculated using the provided FluidProperties UserObject.
+respect to pressure and temperature) will be calculated through interpolation, while all
+remaining (or missing) fluid properties will be calculated using the provided `FluidProperties` `UserObject`.
```
`FluidProperties` object. Same for occurrences below.

GiudGiud replied:
should have caught all occurences

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1176231/) on 629fa44 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22060/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 74c716e338313f2cdcf369c277d6693329c5e102`

joshuahansel commented:
@GiudGiud Why are there so many clang format failures? Are you using the pre-commit hook?

GiudGiud commented:
nope

joshuahansel commented:
You should fix your life by doing this: `./scripts/install_format_hook.sh`. It's what the cool kids do.

GiudGiud commented:
@kyriv1980

GiudGiud commented:
no I m 2 years in Im committed now

joshuahansel commented on modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:
```diff
@@ -474,31 +752,286 @@ TabulatedFluidProperties::s_from_p_T(Real pressure, Real temperature) const
+  p_T_from_v_e(v, e, p0, T0, p, T, conversion_succeeded);
+  const Real s = s_from_p_T(p, T);
+  const Real h = h_from_p_T(p, T);
+  // this is simple the definition of Gibbs free energy
```
Delete

joshuahansel commented:
I don't have any more comments on the fluid properties stuff; I'm "done".

GiudGiud commented on modules/fluid_properties/src/userobjects/SinglePhaseFluidProperties.C:
```diff
@@ -48,6 +70,31 @@ SinglePhaseFluidProperties::s_from_p_T(Real p, Real T, Real & s, Real & ds_dp, R
+  bool conversion_succeeded = true;
+  p_T_from_v_e(v, e, p0, T0, p, T, conversion_succeeded);
+  s = s_from_p_T(p, T);
+  ds_dv = p / T;
+  ds_de = 1 / T;
```
Clausius:
dh = Tds + vdP
then
du = dh - vdP - Pdv
so 
ds = (du + Pdv) / T

GiudGiud committed:
Address Josh's review
- simplify tabulated ve test
- fix units in fluid properties material
- improve formatting of TFP documentation
- dont refer to base class name in context of input files

Add SFP documentation page. It is rather reduced as the main index.md page is mostly SFP material
refs #20101

GiudGiud commented:
should be ready again

GiudGiud committed:
Add parameter groups to TabulatedFluidProperties, refs #20101

GiudGiud commented:
Test failures seem unrelated

joshuahansel commented on modules/fluid_properties/doc/content/utils/FluidPropertiesUtils.md:
```diff
@@ -0,0 +1,45 @@
+
+The 2D Newton method solver is used to solve the problem:
+
+\begin{equation}
+\begin{split}
+$f1(x, y) &= a \\ 
+$f2(x, y) &= b
+\end{split}
+\end{equation}
```
This didn't render for some reason. Is it the $ at the beginning of the lines?

joshuahansel commented on modules/fluid_properties/doc/content/utils/FluidPropertiesUtils.md:
```diff
@@ -0,0 +1,45 @@
+The functions should compute both their return value and the derivatives of their value with regards to
+both variables using references.
+
+The routine will **not** error if a NaN is encountered or if the maximum number of iterations is exceeded.
```
Double * doesn't render in MOOSE-flavored markdown - you need + instead

joshuahansel commented on modules/fluid_properties/doc/content/utils/FluidPropertiesUtils.md:
```diff
@@ -0,0 +1,45 @@
+Instead, a `converged` boolean will be set to false, and the routine will return the last valid
+estimates of $x$ and $y$. It is the responsibility of the user to handle this unconverged output.
+
+## Brent method solver
+
+The Brent method is used to find roots of a general function.
```
The Brent method -> Brent's method

joshuahansel commented on modules/fluid_properties/doc/content/utils/FluidPropertiesUtils.md:
```diff
@@ -0,0 +1,45 @@
+The Brent method is used to find roots of a general function.
+It was adapted from `Numerical Recipes in C++`.
+
+More information may be found [here](https://en.wikipedia.org/wiki/Brent%27s_method)
```
period

GiudGiud committed:
Address Josh'review
- utils doc formatting
Refs #20101

GiudGiud commented:
@kyriv1980 last chance for reviewing, I'd rather this be merged soon

GiudGiud commented:
Failing tests are unrelated and fixed by this other PR: #22126

GiudGiud commented:
Thanks for the reviews!!!

licharlot commented:
@GiudGiud  Thanks for finalizing this!
